### PR TITLE
[SPARK-6608] [SQL] Makes DataFrame.rdd a lazy val

### PR DESCRIPTION
**Pull request by liancheng**
Before 1.3.0, `SchemaRDD.id` works as a unique identifier of each `SchemaRDD`. In 1.3.0, unlike `SchemaRDD`, `DataFrame` is no longer an RDD, and `DataFrame.rdd` is actually a function which always returns a new RDD instance. Making `DataFrame.rdd` a lazy val should bring the unique identifier back.

[Review on Reviewable](https://reviewable.io/reviews/apache/spark/5265)

### CODE REVIEW

**Comment by liancheng** [comment]
@rxin Is there a good reason that makes `DataFrame.rdd` have to be a function?

**Comment by petro-rudenko** [comment]
+1 for this, since for example [the caching logic from ml package](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala#L64) doesn't work properly.

**Comment by liancheng** [comment]
@petro-rudenko Oops, that's a good catch!

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 1]
```diff
@@ -941,10 +941,10 @@ class DataFrame private[sql](
   /////////////////////////////////////////////////////////////////////////////
 
   /**
-   * Returns the content of the [[DataFrame]] as an [[RDD]] of [[Row]]s.
+   * Represents the content of the [[DataFrame]] as an [[RDD]] of [[Row]]s.
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: can you update the doc to say that the RDD is memoized, i.e. once called, even if you change the spark sql configuration, it won't change the plan anymore?

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: Thanks, that's a good point.

**Comment by rxin** [comment]
LGTM.

**Comment by liancheng** [comment]
Merged to master. Thanks for the review!

**Pull request closed without merging**